### DISCUSSION TITLE
Shell integration

### DISCUSSION THREAD

**Discussion by KaruroChori** [saghul/txiki.js#515] [Ideas]
To put it briefly, [bun sh](https://bun.sh/docs/runtime/shell) or [google zx](https://github.com/google/zx) but in txiki.js.

By the way, I tried to check if zx was usable with some minor changes to use the txiki interface, but their entire basecode is incompatible due to https://github.com/saghul/txiki.js/issues/511#issuecomment-2096231405.
If that fix is merged, we might have this functionality more or less for free patching the google implementation a bit.

**Comment by saghul** [comment]
I see you made https://github.com/quickjs-ng/quickjs/pull/402 thank you!

I agree that would be very useful to have.

If you can make `zx` work by patching it, I'm game for including it!

I'll bump the QuickJS version when the PR lands.

**Reply by KaruroChori** [reply]
I have an updated branch and everything set up, so I will be testing it shortly. Hopefully it is going to be an easy ride.

**Reply by saghul** [reply]
This is awesome!

**Reply by KaruroChori** [reply]
The binary I got by updating to master quickjs is not working properly. Trying to run it will result in:
> /archive/shared/projects/txiki.js/src/vm.c:227:TJS_NewRuntimeInternal: Assertion `(qrt->rt) != ((void *)0)' failed.

Was there any change in the APIs which is not reflected in the current txiki code? I am afraid this one is not something I can solve alone :D.

**Reply by KaruroChori** [reply]
Broken after e5ae6cf

**Reply by saghul** [reply]
I'll take a look!

**Reply by saghul** [reply]
https://github.com/saghul/txiki.js/pull/519 a change I made in the way memory limits are handled needed an adjustment.

**Reply by KaruroChori** [reply]
Ok, so the one dependency which was preventing lots of packages from working was fast-glob because of its usage of "reserved" words as names for attributes in classes. With the new cherrypick on quickjs master it is now working.

What we are left with is just patching zx. These are the node dependencies it has, for which alternatives are needed:
```
import assert from 'tjs:assert'
import { spawn, spawnSync, StdioOptions, IOType } from 'node:child_process'
import { AsyncHook, AsyncLocalStorage, createHook } from 'node:async_hooks'
import { Readable, Writable } from 'node:stream'
import { inspect } from 'node:util'
import { basename, dirname, extname, join, resolve } from 'node:path'
import url from 'node:url'
import assert from 'node:assert'
import { createInterface } from 'node:readline'
export { default as path } from 'node:path'
export * as os from 'node:os'
```
Some of them will be easier than others I guess :D.

**Reply by saghul** [reply]
Indeed! readline might be tricky, not sure if it's actually used for core functionality.

**Reply by KaruroChori** [reply]
In total, this is how the different libraries are being used:
```
os.tmpdir, os.homedir
path.join
assert
readline.createInterface
child_process.spawn, child_process.spawnSync, child_process.StdioOptions, child_process.IOType
async_hooks.AsyncHook, async_hooks.LocalStorage, async_hooks.createHook
stream.Readable, stream.Writable
inspect
```
I consider the repl functionality to be removed. Too hard to implement and not that useful.
I might have to cut few more things to make this feasible.

**Reply by KaruroChori** [reply]
Yes readline and properly mapping spawn will take a bit. I have no clue on how hooks are being used. For the stream we should be covered. Inspect can be dropped. Assert we just need to map it in the native module of tjs. And for os and path since they are only barely used, their functionality should be "easy" as well.
